### ISSUE TITLE
ClrMD 2.0 feedback

### ISSUE THREAD

**Issue by leculver** [microsoft/clrmd#311]
I am planning to ship a ClrMD 2.0, which will allow me to make significant changes to the library.  I have put the [ClrMD v2 roadmap here](https://github.com/microsoft/clrmd/blob/master/doc/ReleaseNotes2.0.md) that explains what the overall plan is.

Right now the document is a bit sparse because I am looking for feedback from the community before getting started.  Please use this issue to give any feedback, questions, concerns, criticisms, etc of the plan of record.

There are more details there, but at its core moving your project to ClrMD 2.0 will require code changes since we will be making significant breaking changes, a lot of cleanup, and refactoring most of the codebase.  With that said, the same basic surface area (with some changes) will exist, so it should be a reasonable amount of work to port from 1.0 to 2.0 in your projects, if you choose to do so.

For those who don't wish to migrate, I still plan to fix critical issues in ClrMD 1.0 for the foreseeable future.  However, I plan to _only_ do new feature work in the ClrMD 2.0 codebase, but I would be happy to accept pull requests from the community who want to backport features from 2.0 -> 1.0.  In theory, I'm hoping this is the best of both worlds for everyone.

What does everyone think?  Are there any concerns or requests you have for ClrMD v2?

**Comment by leculver** [comment]
FYI to some folks who have submitted recent issues or prs:  @mitikov @0xd4d @isakiev @drewnoakes @loop-evgeny.

**Comment by drewnoakes** [comment]
For API changes, #163 comes to mind.

**Comment by leculver** [comment]
@drewnoakes  Yes, that kind of change is absolutely the kind of stuff I'll be doing in ClrMD 2.0.  One of my tasks for this week is to go back through all of the open issues and mark them as 1.0 or 2.0 changes.  That one we'd go fix in ClrMD 2.0.

**Comment by mitikov** [comment]
The roadmap looks doable, 
**Tests**: as long as core is built on abstractions, it shall be testable.
I've never seen/managed to get CPU-bound code with ClrMD; I/O was always the bottleneck for me; putting in different words - performance was good; I'd prefer to trade off/attempt to runaway from struct (ClrOBj) to interface/class to ease mocking.

That would allow to get proper **unit** test coverage so that logic inside library is covered.
At the same time the current approach with memory snapshot collection for integration tests might stay eligible to verify runtime/(hosting environment) assumptions used inside core.

The part I missed a lot is a possibility to get the snapshot/attach/(execute code) on demand.
Use case - as soon as something interesting happens (based on my custom condition), I'd like to capture a memory snapshot.

The current way of doing that is modifying source code to with empty try{throw TimeForSnapshotException()}catch{} + configuring tool to act on the exception.

The approach has 2 drawbacks:
1) Product source code gets diagnostic code that you feel shame for =\
2) That does not work in case issue takes place inside code you do not own.

Should we have a possibility to attach to the process, integration testing would evolve as well.

**Comment by mitikov** [comment]
Do you want to write ClrMD 2.0 on F# ?

**Comment by 0xd4d** [comment]
Consider enabling nullable reference types in CLRMD 2.0.

**Comment by leculver** [comment]
@mitikov:

> Do you want to write ClrMD 2.0 on F# ?

I don't know F# so no.  =)

> I'd prefer to trade off/attempt to runaway from struct (ClrOBj) to interface/class to ease mocking.

Setting aside performance for a moment, one of the biggest issues that folks have using ClrMD at scale is memory constraints on 32 bit processes.  I worry about changing ClrObject to be a class because it would generate so much garbage, and if you store these values anywhere they eat up so much more space.  On the other hand, making them a class would make a lot of things easier, and I'll put it on the list to consider.

> Feature request possibility to get the snapshot/attach/(execute code) on demand.

I've read through this section here, but I am not sure I understand what you are asking for.

@0xd4d:

> Consider enabling nullable reference types in ClrMD 2.0.

I was looking into that as well, I will add it to the roadmap.  I'm 100% on board with nullable reference types for some of the surface area, but I worry it'll be annoying to use ClrMD if I over use it.  At minimum, I do plan to turn on the feature and experiment with how much I like using the library afterwards.

**Comment by chrisnas** [comment]
Features should be the same both on Windows and Linux and this might have an impact at the API level. For example, even though I've "fixed" thread stack access on Linux (https://github.com/microsoft/clrmd/pull/275), I've met new issues that seem to be related to ptrace usage. So it might be necessary to "suspend" the process before iterating on its threads and "resume" it afterward.

**Comment by leculver** [comment]
@chrisnas There's two things going on here:

1.  The Linux code in general needs more work and testing, agreed there.  The current testbed is not great for Linux testing, so it's one of the things I want to tackle that regardless of other work.
2.  Currently the only Live process attach supported for Linux is "Passive".  The underlying goal of "Passive" attach is that ClrMD itself will *not* do anything disruptive to the target process such as suspending threads.  This means it's expected that if you want better data, you should suspend the process (or threads) yourself.  That won't change...the thing we need to implement are the other attach modes of ClrMD which will suspend the process on your behalf.

I may change the attach flags concept in v2 to be a more descriptive method name, like maybe `AttachToProcessAndSuspend` and `AttachToProcessNoSuspend`.  Just a thought though.

**Comment by weltkante** [comment]
> one of the biggest issues that folks have using ClrMD at scale is memory constraints on 32 bit processes. I worry about changing ClrObject to be a class because it would generate so much garbage

As long as you are forced analyzing 32 bit dumps with a 32 bit process the memory footprint is definitely an important factor.

> I've never seen/managed to get CPU-bound code with ClrMD; I/O was always the bottleneck for me

I'm not sure I can follow, once you have loaded a dump into memory you shouldn't be I/O bound anymore. Is this library so bad that it never loads the dump into memory even if the address space is available, or are your usecases just too specialized because you only pass through the data once? I don't think your statement can be generalized.

I've been using clrmd a few years ago to prototype interactive inspection of a dump, I didn't pay attention to how much I/O was happening but I'd expect performance to be dominated by answering the user queries and not by reloading what should already be in RAM.

I don't have a good feeling about changing away from structs, most other .NET Core projects tend to go the opposite way and turn as much as possible into structs and spans to get the best performance. This decision should definitely not be taken lightly, without careful profiling both performance and memory footprint.

**Comment by leculver** [comment]
@weltkante 

> I've never seen/managed to get CPU-bound code with ClrMD; I/O was always the bottleneck for me

I think the use case he's talking about that hits the most IO is inspecting the heap.  If your process inspects the heap in a deep or meaningful way you will be very, very IO bound.

**Comment by ryanmolden** [comment]
I've mentioned these before but thought I would just throw them up here for posterity

1) Have a mode where I can indicate what is most important to me speed or memory usage. The simplest example is if I am trying to process a OOM dump of a 32 bit process, basically every byte of memory in the CLRMD host process is precious and I would actually happily trade-off some data fetching latency by having you NOT cache anything, if it means CLRMD itself consumes (hopefully much) less memory. I realize that isn't always the case and perhaps default could be speed (i.e. enable internal caching) and have this be a knob for 'weirdos' :P

2) Second the call for more interfaces to allow for easier mocking. If I write objects in terms of CLRMD types it makes MY unit testing very difficult as I need a whole real CLRMD object graph spun up for each test. If I could just mock the various bits of the CLRMD stuff I care about for testing MY stuff that would be ideal. I realize this is a potential perf issue. In my current code I just ended up creating that interface layer and writing my code against that, but now the perf problem is there AND I have a maintenance / agility problem with my own code.

3) Consider using native memory where it makes sense. This sounds really weird but I have found when processing LARGE amounts of data where the data structures used are going to be short-lived and the types you are processing are simple (like uint/ulong for addresses) that making data structures that implement say IList etc but are backed by native memory can be a large win in controlling managed memory fragmentation and keeping short-lived items off the LOH. Using fixed ptrs and the like the access characteristics are as good or better than the standard .NET collections but you can deterministically free the memory AND prevent a HUGE array (for instance) from landing on the LOH and hanging around in zombie form forever.

4) Revisit the design of some of the core types. For instance it seems odd to me, that, for instance, EnumerateObjectReferencesWithFields is on ClrType not ClrObject, but it requires an Address, i.e. an object. I mean I get that the type info is needed to understand the fields of the object, but inevitably it just means my code, that is otherwise dealing with ClrObject has to say <someCLRObject>.Type.EnumerateRefsOfObject(someCLRObject.Address). Granted some of this could be simplified with extension methods, but just relocating methods to contexts in which they make the most logical sense would also help. There are numerous examples of 'mis-homed' methods. Nothing egregious, but it also makes discovery via Intellisense a bit harder since if I do someCLRObject. I see no methods that looks like EnumerateRefs, instead I have to do someCLRObject.Type. before I see something akin to what I am looking for.

**Comment by leculver** [comment]
FYI, I have branched the main clrmd repo as [release/1.1](https://github.com/microsoft/clrmd/tree/release/1.1).  I'm beginning work in my repo as branch [v2](https://github.com/leculver/clrmd/tree/v2).  Nothing in my branch will be ready for review any time soon, but I'll update this thread as I work.  Please feel free to unsubscribe here if you want less updates.

**Comment by loop-evgeny** [comment]
> So it might be necessary to "suspend" the process before iterating on its threads and "resume" it afterward.

Please don't do that! Not without the user explicitly calling a method like `SuspendProcess()`, at least. What makes ClrMd so useful right now is that I can use it to get stack traces of an application running in production. Suspending it would not be acceptable, so I would no longer be able to use ClrMd.

**Comment by loop-evgeny** [comment]
The roadmap sounds good. My main request is a non-functional one: please take care to provide good error messages and, more generally, a good user (developer) experience. For example, instead of something like "dac architecture mismatch", say "Tried to load the DAC at /usr/lib/... which is a 32-bit DAC, but the current process is 64-bit. The bitness of the ClrMd process must match the bitness of the process/dump you are trying to analyze." It only takes an extra minute or so to write all this, but will save your users so much time!

**Comment by weltkante** [comment]
> What makes ClrMd so useful right now is that I can use it to get stack traces of an application running in production.

Doesn't work reliably and in fact is [already not supported now](https://github.com/microsoft/clrmd/issues/303#issuecomment-556508640), I was looking into that technique to get stacktraces of livelocked threads myself. If the stack of a thread changes while you are enumerating frames you'll get nonsense stacktraces.

I don't know what you're doing so maybe you don't care about being able to diagnose every failure of your service and are just sampling stack traces or something similar, but if you need reliable logging you may want to look into different techniques.

I'm posting mainly to warn future readers who are not aware, not for arguing against, I agree that the passive mode is a nice feature and worth to keep.

**Comment by loop-evgeny** [comment]
Thanks for the heads up. Good to know it's not officially supported! However, it works well enough for my purposes. Even if the results are not 100% accurate it's still very useful, because I don't have any other tool that can shown me (even approximate) stack traces without suspending the process.

**Comment by leculver** [comment]
@weltkante @loop-evgeny:

> Please don't do that! Not without the user explicitly calling a method like `SuspendProcess()`, at least.

I think there's some confusion here, I am not removing any functionality.  Let me step back and make sure things are clear:

You can attach to a live process with three different flags.  A [`Passive`](https://github.com/microsoft/clrmd/blob/317c5c4dfd57f256f03944b4c294c64c28b6741f/src/Microsoft.Diagnostics.Runtime/src/Common/AttachFlag.cs#L32)  attach means:  "ClrMD will not suspend the process.  In order to be in a 'supported' scenario it's expected that you suspend the process before you start calling into ClrMD." 

The other two attach modes ([Invasive/NonInvasive](https://github.com/microsoft/clrmd/blob/317c5c4dfd57f256f03944b4c294c64c28b6741f/src/Microsoft.Diagnostics.Runtime/src/Common/AttachFlag.cs#L12-L23)) have windows-specific meaning, but essentially they mean:  "ClrMD will suspend the process for you, this is fully supported as long as you don't resume the process as ClrMD is working."

Note that only Passive attach is currently implemented for Linux.  In v2 I'll also support suspending the process with the other flags (or more likely different functions).

Here's the important part:  When I say that inspecting a running process is not supported, I only mean that you are stepping outside of the bounds of what the CLR debugging layer (libmscordaccore.so) was designed to handle.  I'm only trying to point out that inspecting a process without suspending it is a risky endeavor.  If you are willing to pay that risk (in weird behavior sometimes) then that's your choice to make.

To be extra clear:

1.  The ability to inspect a running process with ClrMD isn't going anywhere.
2.  "Unsupported" means you might encounter a bug that's literally unfixable.  As long as you are ok with this, then you can use it to your hearts content.
3.  I will fix every bug I that is possible to fix in this scenario, for example I just fixed https://github.com/microsoft/clrmd/issues/303.

Basically, what chrisnas is saying may literally be true:  Changes to the CLR runtime in the future might make it so that inspecting "running" threads stops working in the future even if it works today, and it might not be something that can be fixed at all.  That hasn't happened so far though.

Does that clear things up?

**Comment by Zero3** [comment]
Thanks for asking for feedback :). I have a few bits to add:

1) It would be really nice with some kind of declaration/documentation about which features are stable/safe to use in production environments and which are not.

    For example: In v1, when reading stack traces, I ran into serious issues like https://github.com/microsoft/clrmd/issues/47 (memory leak on every stack trace) and https://github.com/Microsoft/clrmd/issues/49 (exception thrown after updating to latest version of .NET Framework). Thankfully, I ran into these issues before deploying to production, and realized that the feature was too unstable for production use. But it would have been nice if the feature was marked as "unstable"/"not production ready"/"beta"/... so I wouldn't have wasted my time trying to make it work and almost blew up production.

2) What the expected cost of the main operations are. For example, what is the expected time for grabbing a stack trace of a single thread of the current application? Obviously this depends on many factors, but it would be nice with a ballpark figure like 1 ms/10 ms/100 ms/1000 ms/....

3) I personally don't care for backwards compatibility. I would much rather use a solid quality library with breaking changes now and then, than a poor quality library without breaking changes.

**Comment by leculver** [comment]
(Sorry I can't write back to everyone in the thread, if I add a thumbs up to your post it's likely I just agree with most of it and will incorporate that into the plan.)

@Zero3:

> It would be really nice with some kind of declaration/documentation about which features are stable/safe to use in production environments and which are not.

As best I can, yes I agree.  One of my main goals here is to reduce the amount of cruft in the library so I can respond to issues faster and for everyone to make it clear what's expected and not.

The main thing I'd say is that in clrmd 2.0, everything will be supported except for inspecting an unsuspended process.  That scenario will be "supported" to the extent that I'll fix any bugs I can if you want to do that, but I'd _highly_ recommend using CreateSnapshotAndAttach instead, since that's what it's designed for.

As for the issues you point out, #47 was actually a bug in CLR itself...if you run into something there then there's only so much I can do, like fixing it in v.next of that product.  If I can work around the problem I will, but when a library I depend on is failing there's only so much I can do.  #49 is something I'd like to react better to in the future, first by saying that kind of code won't be in the library but second that if the overall surface area of clrmd is smaller it'll be a lot easier to fix these issues as they come up.

> What the expected cost of the main operations are.

There's so many factors here I'm not sure I'll be able to come up with a good answer there...  I could always run some tests and post some averages.

> I personally don't care for backwards compatibility. I would much rather use a solid quality library with breaking changes now and then, than a poor quality library without breaking changes.

This is my philosophy too and why I'm taking this approach.  I'm still keeping everything that makes sense, but for the most part 2.0 will be a clean slate.

**Comment by leculver** [comment]
Another question:  Does anyone here use Desktop .Net 3.5 support?  **Thumbs up on this message to vote to keep supporting .Net 3.5, thumbs down on this message to vote to remove it.**  I'm leaning towards leaving it in the product.

The biggest problem with 3.5 support is that it's almost entirely untested.  Adding tests for it would really increase the test matrix (and also require you to have 3.5 installed on your machine to fully run all tests...unless we come up with another system).

**Comment by leculver** [comment]
After the feedback here and more thought I've removed .Net 3.5 support: https://github.com/microsoft/clrmd/pull/342.

Going forward I'll still maintain support in ClrMD 1.1 for v3.5 and v4.0 RTM versions of CLR.  (Note that CLR 4.5 and beyond are supported.  Very few computers are stuck on CLR v4.)

I've only removed support (not done any cleanup yet) but this change should let me greatly simplify the internals of ClrMD.

**Comment by paulomorgado** [comment]
@raffaeler, your input would be appreciated.

**Comment by raffaeler** [comment]
Thanks @paulomorgado for signalling this thread.
The first thing coming into my mind is that roadmap should mention this [proposal](https://github.com/microsoft/clrmd/issues/302) which has already been tagged for 2.0

In general, I would like to see more of the [SOS](https://github.com/dotnet/diagnostics/tree/master/src/SOS) logic moved into the CLRMD library so that you can either use it as a plugin/tool or as a library to build diagnostics that is app-specific.
Currently the SOS dll just receives strings formatted from the native code, therefore it **cannot** be used as a library (parsing messages is weird) .
Being able to create self-diagnosing capabilities in existent apps is one of my goals.

**Comment by raffaeler** [comment]
With regards to the Portable PDB and ECMA metadata in general, I am successfully using the [System.Reflection.Metadata](https://github.com/dotnet/corefx/tree/master/src/System.Reflection.Metadata) to both read assemblies and read/write pdbs. The library is hard to use (also because of the lack of documentation and weird logic when reading custom attributes) but I would prefer CLRMD use that instead of custom code (unless there are other good reasons that I don't know).

**Comment by leculver** [comment]
FYI if anyone is submitting pull requests right now, I have a draft PR up as https://github.com/microsoft/clrmd/pull/358.  It's not complete or ready for review...  I plan to have the main refactoring done and this PR complete by Dec 5th if all goes well.

**Comment by leculver** [comment]
> Second the call for more interfaces to allow for easier mocking

@ryanmolden (and others):  I've completed the initial pass at refactoring with "easier mocking" in mind.  Here's the main points:

1.  The library is largely uncoupled now.  For example, ClrType implementations don't take a deep dependency on "DesktopHeapType", etc.  As an example, this means you can more easily swap your own implementation of ClrHeap while leaving most of the rest of the runtime alone.
2.  Most abstract Clr* types have a concrete "Clrmd*" implementation, for example, `abstract ClrModule` has a concrete `ClrmdModule` which is its real implementation.  These are marked public so you can build them yourself.
3.  Clrmd* implementations take an IFooData interface which contains the raw data used to build the object (and is not held onto by the object), and IFooData contains an IFooHelpers, which is held onto for lazy evaluation/method implementations.  Here's an example from [ClrmdModule](https://github.com/microsoft/clrmd/blob/testing/2.0/src/Microsoft.Diagnostics.Runtime/src/Implementation/ClrmdModule.cs#L40).
4.  If you want to mock clrmd types, you can use these Clrmd* types and provide your own IFooBuilder/Helpers.

Here's an example of constructing my own `ClrModule` for testing purposes using `ClrmdModule` and my own IModuleData implementation:  https://github.com/leculver/clrmd/commit/3e8b4cf0a2957597bd3c5b125903d8aa0db146b9

This solution isn't perfect, but I would rather avoid having a large amount of IClrModule, IClrType, etc which I can't add methods to...  For now I _think_ this lets you build your own ClrMD types for testing without too much of the complications I am trying to avoid by not adding a bunch of interfaces to represent clrmd constructs.

Let me know if there are questions or concerns, or if this doesn't get you where you need to be.

**Comment by Zhentar** [comment]
Related to the System.Reflection.Metadata request, I'd like if `System.Reflection.PortableExecutable` can be used in place of ClrMd specific PE data structures where possible - it would make it easier for me to use the same code to work with both native & .NET modules.

Also pinging @jazzdelightsme for any thoughts he might have

**Comment by leculver** [comment]
@raffaeler / @Zhentar:

I have filed issues #415 to track using System.Reflection.Metadata in place of IMetaDataImport and PEImage.  The only real concern is whether all of the functionality needed exists in the library, and of course time required to change everything over to it.

@raffaeler:
> In general, I would like to see more of the SOS logic moved into the CLRMD library so that you can either use it as a plugin/tool or as a library to build diagnostics that is app-specific.

Unfortunately that's not quite a design goal of ClrMD itself, but I agree with what you are asking for.  I'm trying to trim down ClrMD to being just the building blocks for other tools and pushing a lot of the uses of ClrMD out of the library.  So things like SOS commands I would expect you to be able to build on top of ClrMD but not something to keep in the library.

I do agree with your overall ask though.  I would really like to see a C# based SOS (or similar library) that exposes a reasonable data model such that you could run !commands in dbgeng that print them to the screen, or load that as a library for your own analysis.

Over the last few years I've sat down to rewrite parts of SOS using ClrMD to see how viable the approach is, and while the results looked very good I simply didn't have the time to follow through on that project.  The `dotnet-analyze` diagnostic feature looked promising from this standpoint but after the initial work to create it, I haven't had a lot of time to dig into that feature to see where things stand.

**Comment by leculver** [comment]
As of this point I _think_ I have either addressed all requests in this thread in the current ClrMD 2.0 codebase, or as issues filed against ClrMD.  If you feel that I have missed a comment here or would like to know how something was fixed/addressed please feel free to file an issue (or reply back here).  This thread is super helpful but since it's so long I might have unintentionally missed something.

Before end of day tomorrow (5PM PST, Dec 13th, 2019) I will finish updating the [https://github.com/microsoft/clrmd/blob/master/doc/ReleaseNotes2.0.md](release notes for ClrMD 2.0 beta), and I'll ship a pre-release beta version of ClrMD 2.0 to NuGet.  Note the API surface area is **not** finalized until I ship a "release" NuGet package without "-beta", so if you want to give it a try and have any feedback or find any problems please do file an issue.  There's still plenty of time for feature requests and API surface changes.

I plan to leave the library in beta until around March or April to really give myself some time to make heavy use of the library and fix any remaining issues with the API surface that will be difficult to fix later.  I _expect_ (but not promising) the API surface to be stable around end of January except for tweaks, so you could probably use it for a real project around then.

Lastly, I wanted to say thanks to everyone who gave feedback and a huge thank you to @NextTurn who has done a ton of work in the codebase for 2.0.  His work has made this rewrite go much faster.

**Comment by leculver** [comment]
Closing this issue now that we are pretty far along in the rewrite.  I'm still monitoring this thread if anyone has more comments or questions.  Thanks all who participated!

**Comment by leculver** [comment]
Thank you to everyone who contributed here.  I've released ClrMD 2.0 today:  https://www.nuget.org/packages/Microsoft.Diagnostics.Runtime/2.0.130507

I'm tracking one major issue that I ran out of time on: #727.  I'll be tackling it first thing Monday.

**Comment by paulomorgado** [comment]
@leculver, is there any announcement of this? Blog post? Docs?

**Comment by leculver** [comment]
@paulomorgado Not yet.  This is more of a soft release.  I'm trying to get #727 wrapped up before I make any big posts about it.  I _think_ that issues is only relegated to a narrow usage of ClrMD but I really want to get it fixed next week before making any big announcement.

It's fine to blog/post about it yourself though.  I just haven't had time to see if CLR wants to do a post about it.

**Comment by holc-tkomp** [comment]
Maybe more of a general feedback, but I'd like to mention that this repo is quite scarce in tags and releases and there was no new tags/releases since Dec 19, 2019 (2.0.0-beta.19618.7).

Mentioning this, as having tags and releases makes it easier to compare what changed between two selected versions, and that's  actually what I'm doing right now - reviewing what changed between 1.0.5 and 2.0.2 and what does it mean for our product.

**Comment by NextTurn** [comment]
> reviewing what changed between 1.0.5 and 2.0.2

v1 and v2 are almost two different products. Did you plan to review changes like this:

![image](https://user-images.githubusercontent.com/45985406/85357853-de49bf00-b544-11ea-9a0b-a45cf62a4ae9.png)

**Comment by holc-tkomp** [comment]
Only a somewhat yes, with different points (`1b71d7e` as 1.0.5 and `009f61a` as 2.0.2), but I don't want to focus on checking every single commit, especially that a lot of the stuff I'm probably interested in, was in PRs for `testing/2.0`, which on master is just one big squashed commit ([Merge testing/2.0 into master (#400)](https://github.com/microsoft/clrmd/commit/b529bfb7a8d69ec1ff51d3cdfd807d6d3895b001)).

As my goal is to check what issues in our project got resolved by the 2.0 version and what new issue/challenges may appear, 1.1 could be good enough comparison point, but since it's possible, using the same version as we're using in the project is better.

Best for me would be detailed change notes, such as
 - Opening 64bit dumps of 32bit processes is now supported (#289)
 - Added `ClrType.EnumerateGenericParameters` (#431)

but I'm aware that this might be hard to make, as probably some fixes required some other stuff to be reverted, etc.. So I'm mostly browsing the release notes, issues, PRs, roadmap(s) to extract what might affect our product (in either positive or negative way). If I find description somewhat interesting, then I lookup the related commits/changes and if they're still in the 2.0.2.

**Comment by raffaeler** [comment]
@holc-tkomp IMO it is far simpler to face the differences one at a time.
I did it in my library and it didn't take much time to make the required changes.
Of course, I may have not used things that you require in your project, so I don't want mine to be a general statement.

I found easier to "navigate" the object model and in fact I build some Linq Extensions that easily allows to query ClrObject, ClrType, etc. to retrieve the needed information.

**Comment by holc-tkomp** [comment]
> I found easier to "navigate" the object model and in fact I build some Linq Extensions that easily allows to query ClrObject, ClrType, etc. to retrieve the needed information.

Same here, unfortunately our product uses some features, that were considered too slow and/or not necessary and/or buggy, thus got removed.

> IMO it is far simpler to face the differences one at a time.
> I did it in my library and it didn't take much time to make the required changes.

I'm trying to do so as well, but while fixing some of the missing type/member due to 1.1->2.0 changes are as simple as changing `ClrRoot` to `IClrRoot`, others are a bit tricky, such as detecting whether the root is a root of a static variable, and for some I haven't found a solution yet - e.g. using `EnumerateMemoryRegions` to collect as detailed as possible information about used, commited, allocated memory in the dump (might create issue for this somewhere later today, to don't spam here).

Dropping support for .NET Framework <4.5 runtimes is also a huge thing to consider for us, as so far we state to support frameworks 2.0 - 4.7.2. So I have to research how long it may take to upgrade from 1.0.5 to 2.0.2, what issues it'll fix, what new issues will it bring, to decide whether in future we'll upgrade to ClrMD 2.0 or maybe implement our own memory dump reader library from scratch.

**Comment by raffaeler** [comment]
> Same here, unfortunately our product uses some features, that were considered too slow and/or not necessary and/or buggy, thus got removed.

It mostly depends on the queries complexities, but if you avoid ToList/ ToArray, you should not suffer from a big overhead. Of course you have to be very careful in grouping as well.

>I'm trying to do so as well, but while fixing some of the missing type/member due to 1.1->2.0 changes ...
If they are not trivial, I would open an issue for each of that clearly marking the issues as migration problems. At least they will remain as suggestions for others as soon as they are resolved either from MS or from the community.

> Dropping support for .NET Framework <4.5 runtimes is also a huge ...

In this case youre best chance is to create your own abstraction which depends on both 1.1 and 2.0 libraries. The manifest of the nuget package can ease the library consumption. Not the perfect solution, but I fully undesrtand the decision from MS to remove that support. The library now is also a bit easier than before.

HTH

**Comment by leculver** [comment]
I realize this is confusing but this issue was originally for planning ClrMD 2.0, the current feedback thread is https://github.com/microsoft/clrmd/issues/743.

I'm locking this to avoid further confusion and will reply back there.
